lesson26/src/js/login.js

```diff
@@ -1,12 +1,55 @@
 import * as validation from "./validation";
 import { togglePassword } from "./module/toggle-password";
+import { Chance } from "chance";
+const chance = new Chance();
 
 const submitButton  = document.querySelector('[data-id="submit_button"]');
+const userNameInput = document.getElementById('username');
+const passwordInput = document.getElementById('current-password');
+const togglePasswordButton = document.getElementById('js-toggle-password');
 
 submitButton.addEventListener("click", (e) => {
     e.preventDefault();
+    login();
 });
 
+const isRegisteredUser = (inputValues) => {
+    const userData = {
+        name:"Nakagawa",
+        email: "nakagawa@sample.com",
+        password: "N302aoe3"
+    }
+
+    return ( inputValues.userName === userData.name || userData.email ) && inputValues.password === userData.password;
+}
+
+const checkRegisteredUser = () => {
+    const inputValues = {
+        userName : userNameInput.value,
+        password : passwordInput.value
+    }
+
+    return new Promise((resolve,reject) => {
+        if( isRegisteredUser(inputValues) ){
+            resolve({ token: chance.apple_token(), ok: true , code: 200 })
+        }else{
+            reject({ ok: false, code: 401 })
+        }
+    }) 
+}
+
+const login = async () => {
+    try{
+        const response = await checkRegisteredUser();
+        if(!response) return;
+        localStorage.setItem("token",JSON.stringify(response.token));
+
+        window.location.href = "./index.html";
+    }catch{
+        window.location.href = "./401.html";
+    }
+}
```

Comment:
IMO
I have received a review that the set of tokens should have separate functions. This is because setting tokens is another task.
For example ↓

```js
const verifyRegisteredUserData = async () => {
  try {
    return await checkRegisteredUser();
  } catch {
    window.location.href = "./401.html";
  }
};

const init = async () => {
  const result = await verifyRegisteredUserData();
  if (result) {
    localStorage.setItem("token", JSON.stringify(response.token));
    window.location.href = "./index.html";
  }
};
```

Reply:
@csakyo 
Thank you for your comment!

Indeed, I thought it would be better. 

Since screen transitions are also a separate task, I separated them into separate functions. 
I would be happy if you could confirm this.
https://github.com/nakagawamai/js-lesson/pull/31/commits/4c33d069116d0ccb21f834769f1e4f4997370956

Reply:
一度修正後に、すみません。
この部分は、下記のようなものだとtokenをセットする部分を、切り分けたことにならないでしょうか？

コードの見通しが悪い感じたので、できれば
resolveの処理の`window.location.href = "./index.html"`と、
rejectoの処理の`window.location.href = "./401.html"`は同じ関数内に記述したいと思いました。

```JavaScript
const setToken = (token) => {
    localStorage.setItem("token",JSON.stringify(token));
}

const login = async () => {
    try{
        const response = await checkRegisteredUser();
        if(response.token){
            setToken(response.token);
        }
        window.location.href = "./index.html"
    }catch{
        window.location.href = "./401.html";
    }
}
```

Reply:
個人的見解になるかもしれず申し訳ないのですが、
tryの中にいくつも処理を書くべきでないと思ってます。tryの中でエラーが起こった場合、その時点で処理は終了してcatchに行くため、どの時点でエラーが起こったかが分かりにくいためです。また、今回で言うと、resolve内の返り値をきちんと受け取れた場合にindex.htmlに遷移できるはずなので、tryの中にその記述もあるべきではないと考えます。

``同じ関数内に記述したい``の自分が思いついたのは下記です。

第一案(init内条件分岐編)
https://codesandbox.io/s/xenodochial-jang-j5nclj?file=/src/js/login.js

第二案(tryの外でif文編)
https://codesandbox.io/s/wonderful-fire-1bq0pd?file=/src/js/login.js

ここの関数については好みの問題かもしれないです...
上のコードは参考程度に見てもらえればと思います:bow:

Reply:
お返事＆解説、ありがとうございます🙏

どの時点でエラーが起こったかが分かりにくいので、tryの中にいろいろ書くべきではない、というご説明に、
とても納得できました。

コードが見通しやすい方がいいのではないか、と思ったのですが、
エラーが発生したときに、エラーがどこで起こったのかがわかりやすいことが大切ですね！

> resolve内の返り値をきちんと受け取れた場合にindex.htmlに遷移できるはずなので、tryの中にその記述もあるべきではないと考えます。

ここの仕様の理解は、足りてなかったと思いました。
確かに言われてみると、その通りだなぁと思います。

ご提案いただいたコードを拝見させていただき、もう少し考えてみようと思います！
いったん現状でマージし、次回のPR時に、変更させていただくかもしれません🙇

Reply:
@csakyo 
たくさんのご提案ありがとうございます！
二つの目のご提案を参考に、下記のように変更してみました。
```JavaScript
const login = async () => {
    let response;
    try {
      response = await checkRegisteredUser();
    } catch {
      window.location.href = "./401.html";
    } finally {
        if (response.token) {
            window.location.href = "./index.html";
            localStorage.setItem("token", JSON.stringify(response.token));
        }
    }
}
```

下記の部分は、Promiseが実行されたら、
Promiseの結果に関わらず、必ず実行されるように、finallyに入れました。

```JavaScript
  // データがきちんと渡ってきて、なおかつtokenが受け取れたらtokenを埋め込み、遷移できる
  if (response.token) {
    window.location.href = "./index.html";
    localStorage.setItem("token", JSON.stringify(response.token));
  }
```